pkg/mutation/system_external_data.go

```diff
@@ -0,0 +1,210 @@
+package mutation
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/open-policy-agent/frameworks/constraint/pkg/apis/externaldata/v1alpha1"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/externaldata"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/types"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// resolvePlaceholders resolves all external data placeholders in the given object.
+func (s *System) resolvePlaceholders(obj *unstructured.Unstructured) error {
+	providerKeys := make(map[string]sets.String)
+	hasPlaceholders := false
+
+	// recurse object to find all existing external data placeholders
+	var recurse func(object interface{})
+	recurse = func(object interface{}) {
+		switch obj := object.(type) {
+		case map[string]interface{}:
+			// our base case - check if this object is an external data placeholder
+			if placeholder, ok := types.ToExternalDataPlaceholder(obj); ok {
```

Comment:
This could potentially be a security hole... if there is an object that has `x-preserve-unknown-fields` set, a requestor could leverage this to send arbitrary external data requests.

If we do not convert placeholders to `map[string]interface{}`, and instead leave them as typed objects, then use type checking to detect placeholders, this shouldn't be an issue, as only placeholders created by the process should be considered.

 We'll also avoid serialization/deserialization costs.